src/client/pipeline.ts

```diff
@@ -20,244 +11,76 @@ declare global {
-  }
-
-  private buildSpeechConfig(config: PipelineConfig) {
-    const speechConfig = config.speechConfig || DefaultSpeechConfig
-    speechConfig.baseKeywordUrl = config.baseUrls?.keyword
-    speechConfig.baseWakewordUrl = config.baseUrls?.wakeword
-    return speechConfig
-  }
-
-  async listen(): Promise<Pipeline> {
-    const stream = await startRecord()
-    if (!stream) {
-      throw new Error(
-        `There was a problem starting the microphone. ${
-          navigator.__polyfilledMediaDevices
-            ? 'This browser does not support microphone access.'
-            : 'Please check the permission and try again.'
-        }`
-      )
-    }
-
-    try {
-      this.audio.source = this.audio.context.createMediaStreamSource(stream)
-    } catch (err) {
-      // This is thrown by Firefox, which will create an `AudioContext`
-      // with a given sample rate, but won't adjust the sample rate of a
-      // user's input device, connect an input stream to an audio context
-      // with a different sample rate, or expose `sampleRate` in the
-      // constraints, capabilities, or settings of the stream's tracks
-      // so you can make a new context before the error happens.
-      //
-      // If this gets fixed, this class can be cleaned up significantly and
-      // the `audio` fields pulled up to the top level.
-      if (err instanceof DOMException && err.name == 'NotSupportedError') {
-        this.audio = this.connectAudio()
-        const sampleRate = this.audio.context.sampleRate
-        if (!sampleRates.includes(sampleRate)) {
-          throw new Error(`unsupported sample rate: ${sampleRate}`)
-        }
-        this.speechConfig.sampleRate = sampleRate
-        this.audio.source = this.audio.context.createMediaStreamSource(stream)
-      } else {
-        throw err
-      }
-    }
-
-    this.audio.vad.connect(this.audio.source, this.audio.processor)
-
-    return new Promise<Pipeline>((resolve) => {
-      this.audio.vad.calibrate(() => {
-        resolve(this)
-      })
-    })
-  }
-
-  private connectAudio(opts = {}): PipelineAudio {
-    this.disconnect()
-
-    const context = new (window.AudioContext || window.webkitAudioContext)(opts)
-    if (context.sampleRate != this.speechConfig.sampleRate) {
-      console.log(`sample rate MISMATCH: ${context.sampleRate}`)
-      this.adjustForSampleRate(this.speechConfig, context.sampleRate)
-    }
+interface KeywordConfig extends BaseConfig {
+  profile: PipelineProfile.Keyword
+  keywordClasses: string[]
+}
 
-    const worker = this.initWorker()
+interface WakewordConfig extends BaseConfig {
+  profile: PipelineProfile.Wakeword
+  keywordClasses?: undefined
+}
 
-    const bufferSize = 512
-    const processor = context.createScriptProcessor(bufferSize, 1, 1)
-    const sendBuffer = this.sendBuffer.bind(this)
-    processor.addEventListener('audioprocess', sendBuffer)
+export type PipelineConfig = KeywordConfig | WakewordConfig
```

Comment:
The design goal is that profiles are a convenience feature (thus `profile` would ideally be an optional key), and that users should be able to put together their own combinations of stages and config values, including adding custom stages. This was achievable in my previous commit, but constraining `PipelineConfig` to be a union type that doesn't include the base removes that ability.

Also, making the various configs interfaces instead of types forces the key uniformity; types don't suffer from this. `WakewordConfig` is actually superfluous since the interfaces/types can't dictate values for the stages they use; that work is done elsewhere.

I don't think I'd gotten all the way to where you'd get an error if you passed a profile name and skipped one of its corresponding config's required keys, but I think it's doable with the type-based approach.

Reply:
Rather than including the base, we could add another config that includes a `stages` property. However, if it's a common case, we should add it to the example app. If not, then maybe we don't need it yet.

Reply:
Also, there's little practical difference between types and interfaces except that interfaces are extensible, which is why I generally prefer interfaces. And if you try to use the keyword profile without setting `keywordClasses` as well, you will get a type error.

Reply:
Oh and I don't see how `WakewordConfig` is superfluous. It's necessary because the profile maps to a stages value.

Reply:
> maybe we don't need it yet
True. This would be for the future, when there are more stages to choose from and we've had time to document how to write your own.

> interfaces are extensible
My takeaway from the docs is that types are extensible (my `KeywordProfile` extended `BaseProfile` and added a new field), but you can't reopen the _same_ type and add new fields to it, which is actually what we want here. I was still working on the type error piece, but I _think_ it's possible (I might be mistaken).

>  the profile maps to a stages value
I see what you mean here. My original thought was that `WakewordConfig` doesn't participate in the stage list mapping; it's a value of `PipelineProfile` pointing to a list of stages. What it actually does, though, is provide the type error you mentioned earlier. I think we could get that another way, but I'm not positive.

All of my hot takes here are colored by the work on the other platforms, and I'm admittedly trying to put things together from the top down rather than adding things as they're needed. It's totally reasonable to start out small and grow as we have time, so I'll follow your lead on that. Were there any other cleanup items on your punch list besides doc/tests?

Reply:
I went ahead and pushed a version that includes `stages` as a property in the config. If `stages` is defined, `profile` should not be passed. I *think* this takes care of the concerns you had. I just want to reiterate that using interfaces does not force key uniformity, but we want it whether you use types or interfaces so that you don't have to recast the config later.

Reply:
Nice update; I like the way this fits together. By "key uniformity", I was just talking about the bit where `keywordClasses` had to also be present in `WakewordConfig` as an optional key. This new type scheme does away with even that, though, so great.

Reply:
> had to also be present in WakewordConfig as an optional key

Right, but this would also have been true of types since we are using a union to combine them. If a key was missing on one of the types, you'd get a type error when using the property (unless you cast using `as`).

Reply:
Ah, OK. I don't think I'd run into that problem yet using the types, so I wasn't aware of it. Think that answers all my questions.